### PR TITLE
Add scalar `persist()` operator, #438

### PR DESCRIPTION
**Pull request by MingweiSamuel**

### CODE REVIEW

File: hydroflow_lang/src/graph/ops/persist.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+/// input_send.send(("hello", "oakland")).unwrap();
+/// input_send.send(("hello", "berkeley")).unwrap();
+/// input_send.send(("hello", "san francisco")).unwrap();
+/// flow.run_available();
+/// // ("hello", "oakland, berkeley, san francisco, ")
+///
+/// input_send.send(("hello", "palo alto")).unwrap();
+/// flow.run_available();
+/// // ("hello", "palo alto, ")
+/// ```
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/ops/persist.rs:62]: TODO update docs

File: hydroflow_lang/src/graph/ops/persist.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    persistence_args: &(0..=1),
+    type_args: RANGE_0,
+    is_external_input: false,
+    ports_inn: None,
+    ports_out: None,
+    properties: FlowProperties {
+        // TODO(mingwei): Are these right?
+        deterministic: FlowPropertyVal::Preserve,
+        monotonic: FlowPropertyVal::Yes,
+        inconsistency_tainted: false,
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/ops/persist.rs:80]: @jhellerstein need to take a look at this

File: hydroflow/tests/surface_persist.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+use hydroflow::util::collect_ready;
+use hydroflow::{assert_graphvis_snapshots, hydroflow_syntax};
+
+#[multiplatform_test]
+pub fn test_persist_basic() {
+    let (result_send, mut result_recv) = hydroflow::util::unbounded_channel::<u32>();
+
+    let mut hf = hydroflow_syntax! {
+        repeat_iter([1])
```
Comment (jhellerstein) [inline hydroflow/tests/surface_persist.rs:11]: surprised by the use of `repeat_iter` before `persist`. Seems like an anti-pattern.

File: hydroflow/tests/surface_persist.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    );
+}
+
+#[multiplatform_test]
+pub fn test_persist_pull() {
+    let (result_send, mut result_recv) = hydroflow::util::unbounded_channel::<u32>();
+
+    let mut hf = hydroflow_syntax! {
+        // Structured to ensure `persist()` is pull-based.
+        repeat_iter([1]) -> m0;
```
Comment (jhellerstein) [inline hydroflow/tests/surface_persist.rs:34]: again, why `repeat_iter`? What are you testing for here?

File: hydroflow/tests/surface_persist.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        &*collect_ready::<Vec<_>, _>(&mut result_recv)
+    );
+}
+
+#[multiplatform_test]
+pub fn test_persist_push() {
+    let (result_send, mut result_recv) = hydroflow::util::unbounded_channel::<u32>();
+
+    let mut hf = hydroflow_syntax! {
+        t0 = repeat_iter([1]) -> tee();
```
Comment (jhellerstein) [inline hydroflow/tests/surface_persist.rs:59]: and again, `repeat_iter`

File: hydroflow_lang/src/graph/ops/persist.rs (hunk 2 above)
Comment (jhellerstein) [reply hydroflow_lang/src/graph/ops/persist.rs:80]: Concur.

**Review by jhellerstein — APPROVED** [review]
Code LGTM. I'd like to understand the motivation of the tests better -- comments would help.

**Review by jhellerstein — COMMENTED** [review]
Must fix docs before finalizing PR!

File: hydroflow/tests/surface_persist.rs (hunk 3 above)
Comment (MingweiSamuel) [reply hydroflow/tests/surface_persist.rs:11]: useful in tests, but I'll update the doc example to not use `repeat_iter`

**Pull request merged**